src/index/LocatedTriples.cpp

```diff
@@ -0,0 +1,240 @@
+    }
+  };
+  auto writeTripleToResult = [&numIndexColumns, &result](auto resultEntry,
+                                                         auto& locatedTriple) {
+    for (size_t i = 0; i < numIndexColumns; i++) {
+      (*resultEntry)[i] = locatedTriple->triple_[3 - numIndexColumns + i];
+    }
+    // Write UNDEF to any additional columns.
+    for (size_t i = numIndexColumns; i < result.numColumns(); i++) {
+      (*resultEntry)[i] = ValueId::makeUndefined();
+    }
+  };
+
+  for (auto row : block) {
+    // Append triples that are marked for insertion at this `rowIndex` to the
+    // result.
+    while (locatedTriple != locatedTriples.end() && cmpLt(locatedTriple, row)) {
+      if (locatedTriple->shouldTripleExist_ == true) {
+        // Adding an inserted Triple between two triples.
+        writeTripleToResult(resultEntry, locatedTriple);
+        ++resultEntry;
+        ++locatedTriple;
+      } else {
+        // Deletion of a triple that does not exist in the index has no effect.
+        ++locatedTriple;
+      }
+    }
+
+    // Append the triple at this position to the result if and only if it is not
+    // marked for deletion and matches (also skip it if it does not match).
+    bool deleteThisEntry = false;
+    if (locatedTriple != locatedTriples.end() && cmpEq(locatedTriple, row)) {
+      if (locatedTriple->shouldTripleExist_ == false) {
+        // A deletion of a triple that exists in the index.
+        deleteThisEntry = true;
+
+        ++locatedTriple;
+      } else {
+        // An insertion of a triple that already exists in the index has no
+        // effect.
+        ++locatedTriple;
+      }
+    }
+    if (!deleteThisEntry) {
+      *resultEntry++ = row;
+    }
+  }
+  while (locatedTriple != locatedTriples.end() &&
+         locatedTriple->shouldTripleExist_ == true) {
+    writeTripleToResult(resultEntry, locatedTriple);
+    ++resultEntry;
+    ++locatedTriple;
+  }
+
+  // Return the number of rows written to `result`.
+  return resultEntry - (result.begin() + offsetInResult);
+}
+
+// ____________________________________________________________________________
+std::vector<LocatedTriples::iterator> LocatedTriplesPerBlock::add(
+    const std::vector<LocatedTriple>& locatedTriples) {
+  std::vector<LocatedTriples::iterator> handles;
+  handles.reserve(locatedTriples.size());
+  for (auto triple : locatedTriples) {
+    LocatedTriples& locatedTriples = map_[triple.blockIndex_];
+    auto [handle, wasInserted] = locatedTriples.emplace(triple);
+    AD_CORRECTNESS_CHECK(wasInserted == true);
+    AD_CORRECTNESS_CHECK(handle != locatedTriples.end());
+    ++numTriples_;
+    handles.emplace_back(handle);
+  }
+  return handles;
+};
+
+// ____________________________________________________________________________
+void LocatedTriplesPerBlock::erase(size_t blockIndex,
+                                   LocatedTriples::iterator iter) {
+  AD_CONTRACT_CHECK(map_.contains(blockIndex));
```

Comment:
In general, you can reduce the multiple times hashing using
map.find()  and then working with that iterator.